orangecontrib/prototypes/widgets/owexplpred.py

```diff
@@ -0,0 +1,367 @@
+        inst2 = copy.deepcopy(tiled_inst)
+        iterations_reached = np.zeros((1, noAtr))
+
+        while not(all(iterations_reached[0,:] > self.maxIter)):
+            if not(any(iterations_reached[0,:] > self.maxIter)):
+                a = np.random.choice(atr_indices[0], p=(var[0,:]/(np.sum(var[0,:]))))
+            else:
+                a = np.argmin(iterations_reached[0,:])
+
+            perm = np.random.choice([True, False], batchMxSize, replace=True)
+            perm = np.reshape(perm, (self.batchSize, noAtr))
+            rand_data = self.data.X[random.sample(
+                range(dataRows), k=self.batchSize), :]
+            inst1.X = np.copy(tiled_inst.X)
+            inst1.X[perm] = rand_data[perm]
+            inst2.X = np.copy(inst1.X)
+
+            inst1.X[:, a] = tiled_inst.X[:, a]
+            inst2.X[:, a] = rand_data[:, a]
+            f1 = self._get_predictions(inst1, classValue)
+            f2 = self._get_predictions(inst2, classValue)
+            
+            diff = np.sum(f1 - f2)
+            expl[0, a] += diff
+
+            # update variance
+            steps[0, a] += self.batchSize
+            iterations_reached[0,a] += self.batchSize
+            d = diff - mu[0, a]
+            mu[0, a] += d/steps[0, a]
+            M2[0, a] += d*(diff - mu[0, a])
+            var[0, a] = M2[0, a] / (steps[0, a]-1)
+
+            if (callback()):
+                break
+
+            # exclude from sampling if necessary
+            neededIter = zSq * var[0, a] / (self.error**2)    
+            if (neededIter <= steps[0, a]) and (steps[0,a] >= self.minIter) or (steps[0, a] > self.maxIter):
+                iterations_reached[0,a] = self.maxIter + 1
+                
+                    
+        expl[0,:] = expl[0,:]/steps[0,:]
+
+        #creating return array
+        domain = Domain([self.atr_names], [ContinuousVariable('contributions')])
+        table = Table.from_list(domain, np.asarray(self.atr_names.values).reshape(-1, 1))
+        ordered = np.argsort(np.abs(expl[0]))[::-1]
+        table.Y = expl.T[ordered]
+        table.X = table.X[ordered]
+        return classValue, table
+
+    def _get_predictions(self, inst, classValue):
+        if isinstance(self.data.domain.class_vars[0], ContinuousVariable):
+            #regression
+            return self.model(inst)
+        else:
+            #classification
+            predictions =  (self.model(inst) == classValue) * 1
+            #return self.model(inst, Model.ValueProbs)[1][:,int(classValue)]
+            return predictions 
+
+
+
+class OWExplainPred(OWWidget):
+
+    name = "Explain Predictions"
+    description = "Calculates attribute contributions to final prediction"
+    #icon = "iconImage.png"
+    priority = 200
+    error = settings.Setting(5)
+    p_val = settings.Setting(5)
+
+    class Inputs:
+        data = Input("Data", Table, default = True)
+        model = Input("Model", Model, multiple = False)
```

Comment:
Can fix it all in PyCharm with _Edit > Select All_, followed by _Code > Reformat Code_.